rest_framework_json_api/backends.py

```diff
@@ -0,0 +1,148 @@
+from django.db.models.sql.constants import ORDER_PATTERN
+from django_filters.rest_framework import DjangoFilterBackend
+from rest_framework.exceptions import ValidationError
+from rest_framework.filters import (BaseFilterBackend, OrderingFilter,
+                                    SearchFilter)
+from rest_framework.settings import api_settings
+
+
+class JsonApiFilterMixin(object):
+    """
+    class to share data among filtering backends
+    """
+    jsonapi_query_keywords = ('sort', 'filter', 'fields', 'page', 'include')
+    search_param = api_settings.SEARCH_PARAM
+    ordering_param = 'sort'
+
+    def __init__(self):
+        self.filter_keys = []
+
+
+class JsonApiQueryValidationFilter(JsonApiFilterMixin, BaseFilterBackend):
+    """
```

Comment:
Not quite sure how to get this right. Especially when an API designer wants to define additional query params with different naming per view it would also need to be configurable.

All would be possible to do somehow but I kind of think it makes the life more difficult of an api designer and the gain is very little in the default case.

Wouldn't it be good enough just to verify the params in the other filter backend classes whether they exists and leave this filter class out for now?

From a maintenance perspective I would prefer not to provide this for now.

Reply:
It seems inappropriate to verify the params in other classes as they only "own" their own params.  Verification is already happening there. For instance, `filter[blah]` checks to see if `blah` is a defined filter.
Some of the params are not even backend filters, like `include` which is elsewhere in the serializer code.

The problem this class addresses is that a client would inadvertently misspell a param and then (if they don't notice that) wonder why the action they requested isn't happening. I'm all for making it easier on the client developer to notice these kinds of errors. If they want to "break" JSON:API and add additional parameters they can:
1. Not include this backend filter at all and the default silent behavior of not complaining about params that are not recognized is achieved -- which was kind of the reason for the `ignore_bad_sort_fields` as well; If you really want your API to be silent to all query parameter errors, this is needed.
2. Extend this backend filter to add additional params to the list by changing the `jsonapi_query_keywords` attribute (or add a `JSON_API_QUERY_KEYWORDS` to settings although I've currently left that out to make it difficult to break the standard.)

Needless to say I feel pretty strongly that this is a useful backend filter to help achieve [JSON:API compliance for query parameter handling](http://jsonapi.org/format/#query-parameters):

"If a server encounters a query parameter that does not follow the naming conventions above, and the server does not know how to process it as a query parameter from this specification, it **MUST** return `400 Bad Request`."

Reply:
I didn't address the different naming per view part. I believe the attribute is actually defined for the view. It's certainly declared that way. I'm guessing `backend_filters` classes are mixed in with the view class? I need to write unit tests for these permutations to confirm.

Reply:
I agree that this would make DJA better comply with JSON API spec which is one of ours goals of DJA. 

I think though there is simple no hurry to implement this in DJA and I think we shouldn't rush into implementing this without having a proper understanding how to do it well.
Because once its landed in DJA and we will find that it is not a good fit we will have to go through a deprecation cycle to change it which makes maintaining DJA harder and the code more complex.

Another reason why I think this is hard to implement is that query params can be accessed wherever request is accessible. Query params are not only used in filter backends but also in other parts e.g. pagination is one example but I am sure there are others.
Also it kind of feels odd to add validation into a filter backend which should be there for filtering.
Maybe a middleware might be the better spot but I am not sure and would try to experiment with it first to get a clear understanding.

We could think of an experimental area for this. We haven't done this in the past and would need to rethink how to structure it though.

So why do we not just try it out in our projects (I can also think of one of our projects where it could be helpful)? Once we have it running in production and it is solid we can still consider adding it to DJA. What do you think?

Reply:
I'm going to finish up this PR, leaving it marked WIP, and then maybe we can make an experimental branch and merge it there?  We can always rebase later. I have a working example and test case in my local code but wanted to merge in to the example and test cases here so others can try it and provide feedback.

I would argue that **filtering** for valid query parameters is a legit use of a filter;-) But this could be done as a view mixin too. I think middleware is too early in the stack as we haven't even gotten to DRF at that point, whereas backend filters are a DRF layer.

Reply:
@sliverc FYI - I've closed the PR while I keep working on this. Do we want to create an experiment branch so others can test/contribute?

BTW, using DjangoFilterBackend adds a new dependency for `django-filter`. That could be a problem. It's not core Django or DRF but is the top-rated filtering package and the one reference from DRF docs.

Reply:
In terms of `django-filter`. We have the same issue with `django-polymorphic`. This is marked as optional dependency, we can do the same with `django-filter`. We just have to make sure that DJA can still be used when this dependency is not installed - meaning we have to be careful with imports.

Actually `django-filter` has the exact same issue just the other way around as it is also usable without Django REST Framework. All rest_framework related classes are part of the `rest_framework` module. We could create a `django_filters` module in our case.

In terms of how to move forward:
I am an advocate of small PRs as they are simpler to review and there quicker to get merged. 

What I would suggest how we move forward:
1. Create a PR for `JSONAPIOrderingFilter` - this is a fairly straight forward change and can be merged easily
2. Create a PR for `JSONAPIDjangoFilterBackend` with its own module approach
3. Create a PR for `JsonApiQueryValidationFilter` where we can experiment how this works best. (I will comment on the experimental discussion you started later on)

How does this sound?

Reply:
@sliverc you wrote:

> In terms of django-filter. We have the same issue with django-polymorphic. This is marked as optional dependency, we can do the same with django-filter. We just have to make sure that DJA can still be used when this dependency is not installed - meaning we have to be careful with imports.

Actually there are no imports for django-polymorphic. I don't quite understand how this code works: https://github.com/django-json-api/django-rest-framework-json-api/blob/a3205367c8050c978c636cb6b7867b5351c3b2d5/rest_framework_json_api/serializers.py#L215 w.r.t. django-polymorphic.

> Actually django-filter has the exact same issue just the other way around as it is also usable without Django REST Framework. All rest_framework related classes are part of the rest_framework module. We could create a django_filters module in our case.

I don't quite understand what you mean by creating a `django_filters` module. Are you saying submit this as a PR to django-filter? As it stands right now, I'll either have to wrap the `JSONAPIDjangoFilter` implementation with a try/except of `from django_filters.rest_framework import DjangoFilterBackend` or just make `django-filter` a dependency for DJA.  Please advise.

Reply:
No, what I mean is to create a folder django_filters in DJA and add a __init__,py and backends.py to implement a custom backend. (Hint: Simplified each python file is called a module but more details at https://docs.python.org/3.6/tutorial/modules.html)

If a user of DJA doesn't import this newly created `django_filters` module or uses it in settings then he doesn't have to install django_filters.